Forbidden correlated having clause

EmmyMiao87 opened the pull request:
Fixed #3377

The correlated slot ref should be bound by the agg tuple of outer query.
However, the correlated having clause should not be analyzed correctly so the result is incorrect.
For example: 
```
SELECT k1 FROM test GROUP BY k1 HAVING EXISTS(SELECT k1 FROM baseall GROUP BY k1 HAVING SUM(test.k1) = k1);
```
The correlated predicate is not executed.

The limit offset should be rewritten also when there is subquery in having clause.
For example: 
```
select k1, count(*) cnt from test group by k1 having k1 in (select k1 from baseall order by k1 limit 2) order by k1 limit 5 offset 3;
```
The new stmt should has a limit element with offset.

The view which has subquery in having clause should not be created.
The reason is that the toSql function of query stmt is incorrect when there is subquery in having clause.
So the view def sql is incorrect and could not pass the sql parser.
Now, the view def sql use the original sql directly to fix this error.

Change-Id: If24206a8dee94013225fe7fdd30ab1fb034de34f

xy720 commented on fe/src/main/java/org/apache/doris/analysis/BaseViewStmt.java:
```diff
@@ -109,33 +112,37 @@ protected void createColumnAndViewDefs(Analyzer analyzer) throws AnalysisExcepti
+                colRef = "`" + colRef + "`";
+            }
+            String colAlias = colNames.get(i);
+            sb.append(String.format("`%s`.%s AS `%s`", tableName.getTbl(), colRef, colAlias));
```
It is work for function call?
For example, if user input:
"create view test_view (h1, h2) as select function(a, b), col2 from testTbl"
will become
"select function(a, b) as h1, col2 as h2 from (select function(a, b), col2 from testTbl)"

EmmyMiao87 replied:
This change has been removed.

morningman commented on fe/src/main/java/org/apache/doris/qe/ConnectProcessor.java:
```diff
@@ -225,7 +226,13 @@ private void handleQuery() {
         try {
-            return SqlParserUtils.getMultiStmts(parser);
+            List<StatementBase> result = SqlParserUtils.getMultiStmts(parser);
+            String[] originStmtList = originStmt.split(";");
```
There maybe `;` inside the query statement

kangkaisen commented on fe/src/main/java/org/apache/doris/analysis/LimitElement.java:
```diff
@@ -42,9 +42,9 @@ public LimitElement(long limit) {
 
-    public LimitElement(long offset, long limit) {
-        this.offset = offset;
+    public LimitElement(long limit, long offset) {
```
Why do we change the order?

EmmyMiao87 replied:
changed

kangkaisen commented on fe/src/main/java/org/apache/doris/analysis/SelectList.java:
```diff
@@ -98,6 +98,24 @@ public void rewriteExprs(ExprRewriter rewriter, Analyzer analyzer)
+            strBuilder.append("DISTINCT ");
+        }
+        for (int i = 0; i < items.size(); ++i) {
+            // strBuilder.append(selectList.getItems().get(i).toSql());
```
Remove.

kangkaisen commented:
`The reason is that the toSql function of query stmt is incorrect when there is subquery in having clause.`

@EmmyMiao87  Hi, why can't we make the toSql function correct ?

kangkaisen commented:
> `The reason is that the toSql function of query stmt is incorrect when there is subquery in having clause.`
> 
> @EmmyMiao87 Hi, why can't we make the toSql function correct ?

OK, I see.

EmmyMiao87 committed:
Fixed #3377

The correlated slot ref should be bound by the agg tuple of outer query.
However, the correlated having clause should not be analyzed correctly so the result is incorrect.
For example: SELECT k1 FROM test GROUP BY k1 HAVING EXISTS(SELECT k1 FROM baseall GROUP BY k1 HAVING SUM(test.k1) = k1);
The correlated predicate is not executed.

The limit offset should be rewritten also when there is subquery in having clause.
For example: select k1, count(*) cnt from test group by k1 having k1 in (select k1 from baseall order by k1 limit 2) order by k1 limit 5 offset 3;
The new stmt should has a limit element with offset.

Change-Id: I65c361af54803eac4288f8b9f1e0aa4bfc0b3d4c

EmmyMiao87 committed:
Add ut

Change-Id: Id05fb757216d0e2669c01fb0fcdf01309d62d79e

morningman reviewed (approved):
LGTM
